devito/at_controller.py

```diff
@@ -0,0 +1,303 @@
+final_report_path = os.path.join(reports_folder_path, "final_report.txt")
+
+
+# Gets the best block sizes for given kernel if they are in the final_report_path
+# param - int time order - time order of kernel.
+# param - int space_order - space order of kernel.
+# returns - list - best block size. Starting from outer most dimension
+# returns - None - if report does not exist or does not contain required block sizes
+def get_at_block_size(time_order, space_order):
+    global final_report_path
+
+    if not os.path.isfile(final_report_path):
+        return None  # returns none if report for best block sizes does not exist
+
+    with open(final_report_path, 'r') as f:
+        for line in f.readlines():
+
+            if "time" in line or "space" in line:  # ignores the first line of the report file
+                continue
+
+            split = line.split(' ')
+            if int(split[0]) == time_order and int(split[1]) == space_order:  # finds the one we are looking for
+
+                # Splits, converts all block sizes into int and returns
+                return [int(element) for element in split[2].split(',')]
+
+    return None  # returns none if no matching time/space order was found
+
+
+# Gets optimal block size. Currently works only on linux
+# param - tuple|list shape - shape of kernel
+# param - int time order - time order of kernel
+# param - int space_order - space order of kernel
+# returns - int - most optimal size  for the block
+def get_optimal_block_size(shape, time_order, space_order):
+    # list [time_order][space_order] hardcoded number of loads from the roof line at (change if necessary):
+    # https://docs.google.com/spreadsheets/d/1OmvsTftH3uCfYZj-1Lb-5Ji7edrURf0UzzywYaw0-FY/edit?ts=5745964f#gid=0
+    number_of_loads = [[11, 17, 23, 29, 35, 41, 47, 53], [19, 25, 31, 37, 43, 49, 55]]  # works only till space order 16
+
+    # find the cache size in KB
+    process = subprocess.Popen("cat /proc/cpuinfo", shell=True, stdout=subprocess.PIPE, universal_newlines=True)
+    out = process.communicate()[0]
+
+    for line in out.split('\n'):
+        if "cache size" in line:
+            cache_s = int(re.findall('\d+', line)[0])  # gets the first decimal value
+            break
+
+    core_c = multiprocessing.cpu_count()  # number of cores
+    loads_c = number_of_loads[time_order / 2][space_order / 2]  # number of loads
+
+    # ((C size / cores) / (4 * length inner most * kernel loads)
+    optimal_b_size = math.sqrt(((1000 * cache_s) / core_c) / (4 * shape[len(shape) - 1] * loads_c))
+    return int(round(optimal_b_size))  # rounds to the nearest integer
+
+
+# Class responsible for controlling auto tuning process, copying relevant files and collecting data
+class AtController(object):
+
+    def __init__(self):
+        self.time_order = None
+        self.space_order = None
+
+        at_base_path = "%s/isat" % os.getenv("HOME")  # path to base directory of ISAT auto tuning tool
```

Comment:
This path assumption is not safe. As the bare minimum you need an isat-specific flag to set paths.